File: leetcode/algorithms/tree/validate-binary-search-tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isValidBstSubtree(root: TreeNode | null): [boolean,number,number] {
    if(root===null){
        return [true,Number.MAX_SAFE_INTEGER,Number.MIN_SAFE_INTEGER];
    }
    
    const leftBSTResult = isValidBstSubtree(root.left);
    const rightBSTResult = isValidBstSubtree(root.right);
    
    const validity = leftBSTResult[0] && rightBSTResult[0];
    
    
    return [validity && leftBSTResult[2]<root.val && rightBSTResult[1]>root.val,Math.min(rightBSTResult[1],leftBSTResult[1],root.val),Math.max(rightBSTResult[2],leftBSTResult[2],root.val)]
}

function isValidBST(root: TreeNode | null): boolean {
    
    return isValidBstSubtree(root)[0];

};

File: leetcode/algorithms/tree/validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private boolean isValidBSTImpl(TreeNode root,long minVal,long maxVal){
        if(root==null){
            return true;
        }
        
        return root.val<maxVal && root.val>minVal && isValidBSTImpl(root.left,minVal,root.val) && isValidBSTImpl(root.right,root.val,maxVal);
    }
    
    public boolean isValidBST(TreeNode root) {
        
        return isValidBSTImpl(root,((long)Integer.MIN_VALUE) - 1,((long)Integer.MAX_VALUE) + 1);
    }
}